activation/nipost_test.go

```diff
@@ -281,73 +310,194 @@ func TestNIPostBuilder_BuildNIPost(t *testing.T) {
 	assert.Error(err)
 
 	// fail post exec
-	nb = NewNIPostBuilder(minerID, postProvider, poetProvider, poetDb, db, logtest.New(t))
+	nb = NewNIPostBuilder(minerID, postProvider, []PoetProvingServiceClient{poetProver}, poetDb, db, logtest.New(t))
 	poetDb.errOn = false
 	postProvider.setError = false
 	// check that proof ref is not called again
-	nipost, err = nb.BuildNIPost(context.TODO(), &hash, nil)
+	nipost, _, err = nb.BuildNIPost(context.TODO(), &hash, nil)
 	assert.NotNil(nipost)
 	assert.NoError(err)
 
 	assert.Equal(3, postProvider.called)
 	// test state not loading if other challenge provided
-	poetProvider.EXPECT().PoetServiceID(gomock.Any()).Return([]byte{}, nil)
-	poetProvider.EXPECT().Submit(gomock.Any(), gomock.Any()).Return(&types.PoetRound{}, nil)
+	poetProver.EXPECT().PoetServiceID(gomock.Any()).Return([]byte{}, nil)
+	poetProver.EXPECT().Submit(gomock.Any(), gomock.Any()).Return(&types.PoetRound{}, nil)
 	hash2 := types.BytesToHash([]byte("anton1"))
-	nipost, err = nb.BuildNIPost(context.TODO(), &hash2, nil)
+	nipost, _, err = nb.BuildNIPost(context.TODO(), &hash2, nil)
 	assert.Equal(4, postProvider.called)
 
 	assert.NotNil(nipost)
 	assert.NoError(err)
 }
 
+func createMockPoetService(t *testing.T, id []byte) PoetProvingServiceClient {
+	t.Helper()
+	poet := mocks.NewMockPoetProvingServiceClient(gomock.NewController(t))
+	poet.EXPECT().PoetServiceID(gomock.Any()).Times(1).Return(id, nil)
+	poet.EXPECT().Submit(gomock.Any(), gomock.Any()).Times(1).Return(&types.PoetRound{}, nil)
+	return poet
+}
+
+func TestNIPostBuilder_ManyPoETs_DeadlineReached(t *testing.T) {
+	t.Parallel()
+	// Arrange
+	assert := require.New(t)
+	poetDb := NewPoetDb(sql.InMemory(), logtest.New(t))
+	poets := make([]PoetProvingServiceClient, 0, 2)
+	poets = append(poets, createMockPoetService(t, []byte("poet0")))
+	poets = append(poets, createMockPoetService(t, []byte("poet1")))
+	nb := NewNIPostBuilder(minerID, &postSetupProviderMock{}, poets, poetDb, sql.InMemory(), logtest.New(t))
+
+	g, ctx := errgroup.WithContext(context.TODO())
+	resultChan := make(chan struct {
+		*types.NIPost
+		error
+	})
+	deadlineChan := make(chan struct{}, 1)
+
+	challenge := types.BytesToHash([]byte("challenge"))
+	g.Go(func() error {
+		nipost, _, err := nb.BuildNIPost(ctx, &challenge, deadlineChan)
+		resultChan <- struct {
+			*types.NIPost
+			error
+		}{nipost, err}
+		return nil
+	})
+
+	// Act
+	// Store proofMsg only from poet1
+	proofMsg := types.PoetProofMessage{
+		PoetServiceID: []byte("poet1"),
+		PoetProof: types.PoetProof{
+			Members:   [][]byte{challenge[:]},
+			LeafCount: 1,
+		},
+	}
+	ref, err := proofMsg.Ref()
+	assert.NoError(err)
+	poetDb.StoreProof(ref, &proofMsg)
+	time.Sleep(time.Millisecond * 10)
+
+	// Signal that the time is out
+	close(deadlineChan)
+
+	// Verify
+	result := <-resultChan
+	assert.NoError(result.error)
+	assert.Equal(*result.NIPost.Challenge, challenge)
+	proof, err := poetDb.GetProof(result.NIPost.PostMetadata.Challenge)
```

Comment:
with the change above this simplifies to:

```suggestion
	result := <-resultChan
	assert.Equal(challenge, *result.Challenge)
	proof, err := poetDb.GetProof(result.PostMetadata.Challenge)
```